GitUI/CommandsDialogs/FormCommit.ScriptOptionsProvider.cs

```diff
@@ -0,0 +1,41 @@
+using GitUI.ScriptsEngine;
+
+namespace GitUI.CommandsDialogs
+{
+    partial class FormCommit
+    {
+        private class ScriptOptionsProvider : IScriptOptionsProvider
+        {
+            private FormCommit _formCommit;
+            private FileStatusList? _fileStatusList;
+
+            public ScriptOptionsProvider(FormCommit formCommit, FileStatusList? fileStatusList)
+            {
+                _formCommit = formCommit;
+                _fileStatusList = fileStatusList;
+            }
+
+            IReadOnlyList<string> IScriptOptionsProvider.Options { get; } = new[] { "SelectedFiles", "LineNumber" };
+
+            string? IScriptOptionsProvider.GetValue(string option)
+            {
+                switch (option)
+                {
+                    case "SelectedFiles":
+                        IEnumerable<string>? selectedFiles = _fileStatusList?.SelectedItems?.Select(item => _formCommit._fullPathResolver.Resolve(item.Item.Name));
+                        if (selectedFiles is null || !selectedFiles.Any())
+                        {
+                            return null;
+                        }
+
+                        return string.Join(" ", selectedFiles.Select(file => file.QuoteForCommandLine()));
+
+                    case "LineNumber":
+                        return $"{_formCommit.SelectedDiff.CurrentFileLine}";
+                }
+
+                return null;
+            }
+        }
+    }
+}
```

Comment:
There is no need to implement this multiple times:
```suggestion
using GitCommands;
using GitUI.ScriptsEngine;

namespace GitUI.CommandsDialogs;

internal class ScriptOptionsProvider : IScriptOptionsProvider
{
    private const string _selectedFiles = "SelectedFiles";
    private const string _lineNumber = "LineNumber";

    private Func<IEnumerable<string>> _getSelectedFiles;
    private Func<int?> _getCurrentLineNumber;

    public ScriptOptionsProvider(Func<IEnumerable<string>> getSelectedFiles, Func<int?> getCurrentLineNumber)
    {
        _getSelectedFiles = getSelectedFiles;
        _getCurrentLineNumber = getCurrentLineNumber;
    }

    public ScriptOptionsProvider(FileStatusList fileStatusList, FullPathResolver fullPathResolver, Func<int?> getCurrentLineNumber)
        : this(() => fileStatusList.SelectedItems.Select(item => fullPathResolver.Resolve(item.Item.Name).QuoteForCommandLine()), getCurrentLineNumber)
    {
    }

    IReadOnlyList<string> IScriptOptionsProvider.Options { get; } = new[] { _selectedFiles, _lineNumber };

    string? IScriptOptionsProvider.GetValue(string option)
    {
        switch (option)
        {
            case _selectedFiles:
                IEnumerable<string> selectedFiles = _getSelectedFiles();
                return selectedFiles.Any() ? string.Join(" ", selectedFiles) : null;

            case _lineNumber:
                int? line = _getCurrentLineNumber();
                return line is null ? null : line.ToString();

            default:
                throw new NotImplementedException(option);
        }
    }
}
```

Reply:
Your `QuoteForCommandLine` actually contains multiple bugs. At the same time, this shows the importance of unit tests, so I added a few.

For the `ScriptOptionsProvider` refactor, this puts too much responsibility on the lambdas. Note that:
- `RevisionFileTreeControl.ScriptOptionsProvider` is distinctively different from the other two, and is not trivially implemented as a lambda returning `IEnumerable<string>`
- `FormCommit.ScriptOptionsProvider` is used in 3 places, and adding nearly identical lambdas to each would significantly increase code duplication.
- The `_fullPathResolver` part of the mapping is not saved. Only the `QuoteForCommandLine`, which is a tiny part of it. The switch case is simply replaced with method declarations, so no save there.

An intermediate solution might be creating a common ancestor to all three. Then, it would be easier to use virtual functions instead of lambdas. But I'm not sure if that really saves anything. The common part is much smaller than the differing parts. The only real save is having a common `IScriptOptionsProvider.Options`, which could probably be done a different way, if it was absolutely necessary.

Also, absolutely not liking `throw new NotImplementedException(option);`. It would conflict with the use of `{if:Option}`s later. `null` is already an accepted way for an option within `ScriptOptionsParser` to indicate a replacement is not possible.

Reply:
> Your `QuoteForCommandLine` actually contains multiple bugs. At the same time, this shows the importance of unit tests, so I added a few. 

I see. But you have got my point.
Please do not expect that suggestions in review comments are 100% ready to use.
That the sequence of replacements in `EscapeForPosixCommandLine` is crucial, is worth a code comment.

> For the `ScriptOptionsProvider` refactor, this puts too much responsibility on the lambdas.

It avoids bidirectional dependencies and avoids multiple repetitions of the same code. 

>     * `RevisionFileTreeControl.ScriptOptionsProvider` is distinctively different from the other two, and is not trivially implemented as a lambda returning `IEnumerable<string>` 

I do not see any problem:
```
                        if (_revisionFileTreeControl.tvGitTree.SelectedNode?.Tag is not GitItem gitItem || gitItem.ObjectType != GitObjectType.Blob)
                        {
                            return Array.Empty<string>();
                        }

                        return new string[] { _revisionFileTreeControl._fullPathResolver.Resolve(gitItem.FileName) };
```

>     * `FormCommit.ScriptOptionsProvider` is used in 3 places, and adding nearly identical lambdas to each would significantly increase code duplication.

Have you missed the second constructor?

>     * The `_fullPathResolver` part of the mapping is not saved. Only the `QuoteForCommandLine`, which is a tiny part of it.

This was just another hint how to DRY and how to address some of Igor's comments, and not 100% final.
Since the `_fullPathResolver` might change over runtime, another lambda and another constructor should be factored out.

```
    public ScriptOptionsProvider(Func<IEnumerable<string>> getSelectedFiles, Func<FullPathResolver> getFullPathResolver, Func<int?> getCurrentLineNumber)
        : this(() =>
            {
                FullPathResolver fullPathResolver = getFullPathResolver();
                return getSelectedFiles().Select(fileName => fullPathResolver.Resolve(fileName).QuoteForCommandLine());
            },
            getCurrentLineNumber)
    {
    }
```

> The switch case is simply replaced with method declarations, so no save there.

But a separation of concerns! It makes the identical handling of "not applicable" more visible.

> An intermediate solution might be creating a common ancestor to all three. Then, it would be easier to use virtual functions instead of lambdas.

Less classes is definitely better. The lambdas become simpler with the above.

> Also, absolutely not liking `throw new NotImplementedException(option);`. It would conflict with the use of `{if:Option}`s later. `null` is already an accepted way for an option within `ScriptOptionsParser` to indicate a replacement is not possible.

I cannot agree. `IScriptOptionsProvider.GetValue` shall only be called for entries in `IScriptOptionsProvider.Options`.
If it would return `null`, one cannot distinguish whether this was a wrong / unnecessary call or whether a supported option is not applicable at the moment.

Reply:
Took a slightly different approach for the constructors, but managed to merge the classes.

I accept that `IScriptOptionsProvider.GetValue` should only be called for entries in `IScriptOptionsProvider.Options`, but not in the way that it is implemented now, via the `switch`'s `default`. In a future PR, I'll add an option to override one of the default options by specifying it in `IScriptOptionsProvider.Options`. This can be used to allow the branch tree to specify a branch, instead of relying on the revision, which may be the head of multiple branches.